release-22.1: clisqlshell: implement `COPY ... FROM STDIN` for CLI

otan opened the pull request:
Backport 1/1 commits from #79629.

/cc @cockroachdb/release

Release justification: critical improvement to 22.1

---

Resolves #16392

Steps:
* Add a lower level API to lib/pq for use.
* Add some abstraction boundary breakers in `clisqlclient` that allow a
  lower level handling of the COPY protocol.
* Altered the state machine in `clisqlshell` to account for copy.

Release note (cli change): COPY ... FROM STDIN now works from the
cockroach CLI. It is not supported inside transactions.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/79819)
<!-- Reviewable:end -->

knz commented:
i'm not keen on backporting a dependency on a forked lib/pq if we can merge the change upstream

otan commented:
latest lib pq changes introduced something somewhat unstable due to the cancel changes (when I updated last year a bunch of tests failed) so I'd rather use the fork. but maybe I'm misremembering.. I feel the smaller set of changes is safer as a backport so I'd rather do this for now. It

otan commented:
eh ill give it a shot, but if there are failing tests ill revert

otan committed:
clisqlshell: implement `COPY ... FROM STDIN` for CLI

Steps:
* Add a lower level API to lib/pq for use.
* Add some abstraction boundary breakers in `clisqlclient` that allow a
  lower level handling of the COPY protocol.
* Altered the state machine in `clisqlshell` to account for copy.

Release note (cli change): COPY ... FROM STDIN now works from the
cockroach CLI. It is not supported inside transactions.
